### PR TITLE
Fix CodeParrot training script

### PR DESCRIPTION
**Pull request by loubnabnl**
This PR fixes some features in the training script of CodeParrot:

* use `Pytorch` implementation of `AdamW` instead of `transformers` implementation
* add shuffling of the sequences in the batches
* fix error in weight decay for LayerNorm
* change the tracked loss to the average over batches instead of the main worker loss + compute average over accumulated steps manually for wandb/tensorboard plot
* update requirements

cc @lvwerra

### CODE REVIEW

File: examples/research_projects/codeparrot/scripts/arguments.py [hunk 1]
```diff
@@ -26,6 +26,9 @@ class TrainingArguments:
     shuffle_buffer: Optional[int] = field(
         default=1000, metadata={"help": "Size of buffer used to shuffle streaming dataset."}
     )
+    batch_shuffle_buffer: Optional[int] = field(
+        default=100000, metadata={"help": "Size of buffer used to shuffle batch sequences of a streaming dataset."}
+    )
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/arguments.py:31]: maybe we could share the shuffle buffer variable between streaming and batching. what do you think about 10k as a compromise?

File: examples/research_projects/codeparrot/scripts/arguments.py [hunk 2]
```diff
@@ -26,6 +26,9 @@ class TrainingArguments:
     shuffle_buffer: Optional[int] = field(
         default=1000, metadata={"help": "Size of buffer used to shuffle streaming dataset."}
     )
+    batch_shuffle_buffer: Optional[int] = field(
+        default=100000, metadata={"help": "Size of buffer used to shuffle batch sequences of a streaming dataset."}
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/arguments.py:30]: maybe we could share the shuffle buffer variable between streaming and batching. what do you think about 10k as a compromise?

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 3]
```diff
@@ -256,13 +262,15 @@ def get_lr():
...
 t_start = time.time()
+loss_tracking = 0
 for step, batch in enumerate(train_dataloader, start=1):
     if args.resume_from_checkpoint and step < resume_step:
         continue  # we need to skip steps until we reach the resumed step
     loss = model(batch, labels=batch, use_cache=False).loss
-    log_metrics(
-        step, {"lr": get_lr(), "samples": step * samples_per_step, "steps": completed_steps, "loss/train": loss.item()}
-    )
+    avg_loss = loss.repeat(args.train_batch_size)
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:270]: that's technically not the average, yet, right? maybe we can also just combine it with the next line.

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 4]
```diff
@@ -256,13 +262,15 @@ def get_lr():
...
+    avg_loss = accelerator.gather(avg_loss).mean()
+    loss_tracking += avg_loss.item()
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:272]: we could already divide by `args.gradient_accumulation_steps` here so we don't need to do it later on.
```suggestion
    loss_tracking += avg_loss.item() / args.gradient_accumulation_steps
```

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py [hunk 5]
```diff
@@ -272,16 +280,28 @@ def get_lr():
         else:
             accelerator.backward(loss)
     else:
+        loss_tracking = loss_tracking / args.gradient_accumulation_steps
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/codeparrot_training.py:283]: See above.
```suggestion
```

**Review by lvwerra — COMMENTED** [review]
Just left a few minor comment, otherwise looks good to me! 🚀

File: examples/research_projects/codeparrot/scripts/arguments.py (hunk 2 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/arguments.py:30]: yes sure

File: examples/research_projects/codeparrot/scripts/codeparrot_training.py (hunk 3 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/codeparrot_training.py]: resolved in 178ce667e7e05348f09775b301e71b8bf538772e

File: examples/research_projects/codeparrot/scripts/arguments.py (hunk 2 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/arguments.py]: resolved in 9143d4d779e00845bb4dc9729225500a6fb4440e

**Review by lvwerra — APPROVED** [review]
